wagon.py

```diff
@@ -0,0 +1,1014 @@
+        self.proc = proc
+        self.logger = logger
+        self.log_level = log_level
+        self.aggr = ''
+
+    def run(self):
+        while self.proc.poll() is None:
+            output = self.fd.readline()
+            if len(output) > 0:
+                self.aggr += output
+                self.logger.log(self.log_level, output.strip())
+            else:
+                time.sleep(PROCESS_POLLING_INTERVAL)
+
+
+def run(cmd, suppress_errors=False, suppress_output=False):
+    """Executes a command
+    """
+    logger.debug('Executing: {0}...'.format(cmd))
+    pipe = subprocess.PIPE
+    proc = subprocess.Popen(cmd, shell=True, stdout=pipe, stderr=pipe)
+
+    stderr_log_level = logging.NOTSET if suppress_errors else logging.ERROR
+    stdout_log_level = logging.NOTSET if suppress_errors else logging.DEBUG
+
+    stdout_thread = PipeReader(proc.stdout, proc, logger, stdout_log_level)
+    stderr_thread = PipeReader(proc.stderr, proc, logger, stderr_log_level)
+
+    stdout_thread.start()
+    stderr_thread.start()
+
+    while proc.poll() is None:
+        time.sleep(PROCESS_POLLING_INTERVAL)
+
+    stdout_thread.join()
+    stderr_thread.join()
+
+    proc.aggr_stdout = stdout_thread.aggr
+    proc.aggr_stderr = stderr_thread.aggr
+
+    return proc
+
+
+class WagonError(Exception):
+    pass
+
+
+def _construct_wheel_command(wheels_path='package',
+                             excluded_packages=None,
+                             wheel_args=None,
+                             requirement_files=None,
+                             package=None):
+    pip_executable = _get_pip_path(os.environ.get('VIRTUAL_ENV'))
+
+    wheel_cmd = [pip_executable, 'wheel']
+    wheel_cmd.append('--wheel-dir={0}'.format(wheels_path))
+    wheel_cmd.append('--find-links={0}'.format(wheels_path))
+    if wheel_args:
+        wheel_cmd.append(wheel_args)
+
+    if requirement_files:
+        for req_file in requirement_files:
+            wheel_cmd.extend(['-r', req_file])
+    if package:
+        wheel_cmd.append(package)
+    return ' '.join(wheel_cmd)
+
+
+def _get_wheels(wheels_path, excluded_packages, package):
+    wheels = _get_downloaded_wheels(wheels_path)
+    excluded_packages = excluded_packages or []
+    excluded_wheels = []
+    for package in excluded_packages:
+        wheel = _get_wheel_for_package(wheels_path, package)
+        if wheel:
+            excluded_wheels.append(wheel)
+            wheels.remove(wheel)
+            os.remove(os.path.join(wheels_path, wheel))
+        else:
+            logger.warn('Wheel not found for: {0}. Could not exclude.'.format(
+                package))
+    return wheels, excluded_wheels
+
+
+def wheel(package,
+          requirement_files=False,
+          wheels_path='package',
+          excluded_packages=None,
+          wheel_args=None):
+    logger.info('Downloading Wheels for {0}...'.format(package))
+
+    if requirement_files:
+        wheel_command = _construct_wheel_command(
```

Comment:
I tend to prefer kwargs args when there are many arguments. it is a matter of taste though